### DISCUSSION TITLE
Запросы к внешнему серверу без транзакций через глобальные переменные

### DISCUSSION THREAD

**Discussion by an220022** [totumonline/totum-ru-issues-and-discussions#58] [Как это сделать?] [answered]
Существует ряд внешних сервисов с ограничением по количеству запросов, например в секунду. Для выполнения задания крона требуется отправить более 500 запросов. Обращение к сервису осуществляется через функцию getFromScript. Чтобы не превышать лимиты используется следующее решение while(action: $getFromScript; action: $sleep) где sleep: sleep(sec: 0.1) При выполнении задания крона иногда выходит ошибка "Connection timed out", после чего никаких изменений в Тотум не вносится. При неудачном выполнении задания приходится запускать задание повторно, запросы в сервис платные, поэтому они оплачиваются до тех пор, пока задание не выполнится успешно. Как правильно организовать работу задания крона с внешними сервисами при массовых запросах?

**Comment by ttm-alexeypolunin** [comment]
Оказалось, что у нас не готов параметр `ssh: true` для функции `exec`.

Его сделают на этой неделе  и я сделаю пример как через глобальную переменную вне транзакционной модели собрать данные и записать их одним действием.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
### Описанный подход будет работать начиная с 2.3.37!

### [Видео на YouTube —>](https://youtu.be/tXBKkkqgC94)

В тексте:

1. Создаем глобальную переменную вне транзакционной модели

2. Запускаем параллельные процессы используя `ssh:true` функции `exec`

```
= : while(preaction: $glob_var; action: $listrepl)

	glob_var: globVar(name: "list"; value: $rowlist)
		~rowlist: selectRowList(table: $#ntn; field: 'api_call'; order: 'id' desc;)
		
	listrepl: listReplace(list: $rowlist; action: $sleep; action: $exec; value: "value")
		
		sleep: sleep(sec: 0.2)
		exec: exec(code: @p_dublirovanie.h_exec_field; var: "num" = $#value[api_call]; ssh: true)
``` 

3. В параллельном режиме выполняем запросы к удаленному серверу

4. Получив ответ блокируем переменную

5. Перебираем список в переменной и записываем снимая блокировку

```
=: while(preaction: $gfs; action: $write)

write: globVar(name: "list"; value: $value)

value: listReplace(list: $getvar; action: "answer" = $val; value: "value")

		getvar: globVar(name: "list"; block: 10)

	val: if(condition: $#num = $#value[api_call]; then: $gfs; else: $#value[answer])
	
~gfs: getFromScript(uri: "https://demoen.totum.online/Remotes/ssh"; timeout: 30)

```

6. Записываем следующим действием результаты и обнуляем переменную

```
a1= : setListExtended(table: $#ntn; field: 'answer' = @$list[[answer]]; where: 'api_call' = @$list[[api_call]]; log: true)

a2=: globVar(name: "list"; value: $#lc)
```

**Reply by an220022** [reply]
Спасибо за ответ. Именно то, что нужно

**Comment by andlev7** [comment]
[ttm-alexeypolunin](https://github.com/ttm-alexeypolunin) подскажите плз как дополнить данный код в части отправки запроса gfs: getFromScript(uri: "https://demoen.totum.online/Remotes/ssh"; timeout: 30) если нужно выполнять post запрос и каждый раз там другие параметры, например, gfs: getFromScript(uri:  str`"https://api.data.com/v3/task_post"`; posts: $json; headers: $headers) . Получается в $json каждый раз должны быть свои значения, но никак не получается сформировать этот запрос. Если на шаге 2 заполнять данные для  posts в $json, то не отправляется gfs который на шаге 5

**Reply by ttm-alexeypolunin** [reply]
Я не очень помню как там пример сделан, поэтому пока предварительно: 

- в posts передается строка, а не структура тотум, что бы ее сделать исп jsonCreate
- если там запускаются вложенные коды через exec то передайте параметр varr и вызовите его $#var_with_posts (название произвольное)
